driven/storage/database.go

```diff
@@ -130,11 +140,93 @@ func (m *database) start() error {
 		return err
 	}
 
+	//apply recipients data to the new structure
+	err = m.fixRecipientsLegacyData(client, messages, messagesRecipients)
+	if err != nil {
+		return err
+	}
+
 	go m.firebaseConfigurations.Watch(nil)
 
 	return nil
 }
 
+// fix recipients legacy data
+func (m *database) fixRecipientsLegacyData(client *mongo.Client,
+	messagesColl *collectionWrapper, messagesRecipientsColl *collectionWrapper,
+) error {
+	fn := func(sessionContext mongo.SessionContext) error {
+		//start transaction
+		err := sessionContext.StartTransaction()
+		if err != nil {
+			log.Printf("error starting a recipients legacy data fix transaction - %s", err)
+			return err
+		}
+
+		/// check if the data fix has been applied
+		messagesRecipientsCount, err := messagesRecipientsColl.CountDocumentsWithContext(sessionContext, bson.D{})
+		if err != nil {
+			abortTransaction(sessionContext)
+			log.Println("error checking messages count")
+			return err
+		}
+		if messagesRecipientsCount == 0 {
+			log.Printf("recipients legacy data has NOT been applied, messages recipients:%d - applying data fix..", messagesRecipientsCount)
+
+			//load all messages
+			var allMessages []model.Message
+			allMessagesTimeout := time.Minute * time.Duration(2)
+			err := messagesColl.FindWithContextTimeout(sessionContext, bson.D{}, &allMessages, nil, allMessagesTimeout) //long timeout
```

Comment:
I am somewhat concerned that it may not be possible to load all messages into memory given that we are going to be retrieving every message ever sent. Unfortunately I think this will most likely also exceed the mzximum transaction size limit. 

With this in mind, it may be best to query for any messages that contain embedded recipients, paginate the results, and handle the pages one at a time. For each message we should generate the new `MessageRecipient` documents, insert them, then clear the embedded recipients on the `Message`. 

Since it is possible that something could fail part way through, we would need to remove the condition checking if the fix has been applied. Because we will be querying only for messages that don't have embedded recipients, this check should not be necessary as there will be no results from this query once all messages have been migrated.

Let me know if you have any questions or concerns on this. Thanks!

Reply:
I had tested it with 3000 records and it passed. But I agree that it could fail if the amount is huge on prod. I had asked the db team about the messages count on prod one month ago and they were 7,455.
After all I improved this handling of the messages by parts (400 at once). You could have a look.

Thanks for your concern! cc @pmarkhennessy